File: cpp_solutions/1255.cpp
class Solution {
public:
    int maxScore = 0;
    void backtracking(vector<string>& words, int currentIdx, int currentScore, vector<int>& count, vector<int>& score) {
        int n = words.size();
        for (int i = 0; i < 26; ++i) {
            if (count[i] < 0) return;
        }
        maxScore = max(maxScore, currentScore);
        for (int i = currentIdx; i < n; ++i) {
            int accum = 0;
            for (auto& c : words[i]) {
                count[c - 'a']--;
                accum += score[c - 'a'];
            }
            backtracking(words, i + 1, currentScore + accum, count, score);
            for (auto& c : words[i]) {
                count[c - 'a']++;
            }
        }
    }
    int maxScoreWords(vector<string>& words, vector<char>& letters, vector<int>& score) {
        vector<int> count(26, 0);
        for (auto& c : letters) count[c - 'a']++;
        backtracking(words, 0, 0, count, score);
        return maxScore;
    }
};

File: rust_solutions/1255.rs
impl Solution {
    pub fn max_score_words(words: Vec<String>, letters: Vec<char>, score: Vec<i32>) -> i32 {
        let mut count = [0; 26];
        for &c in letters.iter() { 
            count[(c as u8 - b'a') as usize] += 1;
        }
        let mut freqs = vec![[0; 27]; words.len()];
        for (i, s) in words.iter().enumerate() {
            for b in s.bytes() {
                let j = (b - b'a') as usize;
                freqs[i][j] += 1;
                freqs[i][26] += score[j];
            }
        }
        Self::backtracking(0, &mut count, &freqs)
    }
    fn backtracking(i: usize, count: &mut [i32; 26], freqs: &Vec<[i32; 27]>) -> i32 {
        if i == freqs.len() {
            return 0;
        }
        let mut ans = Self::backtracking(i + 1, count, freqs);
        if (0..26).all(|j| count[j] >= freqs[i][j]) {
            let mut acc = 0;
            for j in 0..26 {
                count[j] -= freqs[i][j];
            }
            let ans2 = Self::backtracking(i + 1, count, freqs);
            ans = ans.max(freqs[i][26] + ans2);
            for j in 0..26 {
                count[j] += freqs[i][j];
            }
        }
        ans
    }
}